File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        answer = 0

        def diameter(root):
            nonlocal answer # IMP int is immutable so need to do this in inner function

            if not root:
                return 0
            
            left = diameter(root.left)
            right = diameter(root.right)

            answer = max(answer, left + right) # diameter will have + , left + right

            return 1 + max(left, right) # while return add 1

        diameter(root) # remember to call it
        return answer
        

    

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        
        
        int count = goodnode(root, root.val);

        return count;
        

        
    }

    public int goodnode(TreeNode node, int value)
    {
        if(node == null)
            return 0;
        if(node.val >= value)
            return 1 + goodnode(node.left, node.val) + goodnode(node.right, node.val);
        else
            return goodnode(node.left, value) + goodnode(node.right, value);

    }


}
